File: app/ui/panels/layer_panel.py
from PySide6.QtCore import Qt, Signal, QSize
from PySide6.QtGui import QIcon, QPixmap, QColor, QPainter, QImage, QStandardItemModel, QStandardItem
from PySide6.QtWidgets import (
    QWidget, QVBoxLayout, QHBoxLayout, QLabel, QPushButton, QSlider,
    QComboBox, QListWidget, QListWidgetItem, QLineEdit, QToolButton, QAbstractItemView, QMenu
)
import numpy as np
from app.core.image_document import ImageDocument
from app.core.layer import Layer
from app.utils.image_utils import numpy_to_qpixmap
from app.utils.logger import logger

class LayerItemWidget(QWidget):
    """
    Custom widget item representing a Layer row in the Layer Manager list.
    Displays: Eye (Visibility), Lock, Thumbnail, Name, and Active Indicator.
    """
    visibility_changed = Signal(str, bool)
    lock_changed = Signal(str, bool)
    name_changed = Signal(str, str)

    def __init__(self, layer: Layer, parent=None):
        super().__init__(parent)
        self.layer_id = layer.id
        self._init_ui(layer)

    def _init_ui(self, layer: Layer):
        layout = QHBoxLayout(self)
        layout.setContentsMargins(4, 2, 4, 2)
        layout.setSpacing(6)

        # 1. Eye Visibility Toggle Button
        self.btn_eye = QToolButton(self)
        self.btn_eye.setText("👁" if layer.visible else "🙈")
        self.btn_eye.setFixedSize(26, 26)
        self.btn_eye.setStyleSheet("QToolButton { border: none; background: transparent; font-size: 14px; }")
        self.btn_eye.clicked.connect(self._toggle_visibility)
        layout.addWidget(self.btn_eye)

        # 2. Lock Toggle Button
        self.btn_lock = QToolButton(self)
        self.btn_lock.setText("🔒" if layer.locked else "🔓")
        self.btn_lock.setFixedSize(24, 24)
        self.btn_lock.setStyleSheet("QToolButton { border: none; background: transparent; font-size: 12px; }")
        self.btn_lock.clicked.connect(self._toggle_lock)
        layout.addWidget(self.btn_lock)

        # 3. Layer Thumbnail Preview
        self.lbl_thumb = QLabel(self)
        self.lbl_thumb.setFixedSize(32, 32)
        self.lbl_thumb.setStyleSheet("QLabel { background: #2b2b2b; border: 1px solid #444; border-radius: 4px; }")
        self.update_thumbnail(layer)
        layout.addWidget(self.lbl_thumb)

        # 4. Layer Name (Editable on double-click or inline)
        self.lbl_name = QLabel(layer.name, self)
        self.lbl_name.setStyleSheet("QLabel { font-weight: bold; color: #e0e0e0; }")
        layout.addWidget(self.lbl_name, stretch=1)

        # Inline LineEdit for renaming (hidden by default)
        self.txt_name = QLineEdit(layer.name, self)
        self.txt_name.hide()
        self.txt_name.editingFinished.connect(self._finish_rename)
        layout.addWidget(self.txt_name, stretch=1)

    def update_thumbnail(self, layer: Layer):
        """Generates a 32x32 thumbnail preview of the layer image."""
        if layer.image is not None:
            h, w = layer.image.shape[:2]
            if layer.mask is not None:
                rgba = np.dstack((layer.image, layer.mask))
            else:
                alpha = np.full((h, w, 1), 255, dtype=np.uint8)
                rgba = np.dstack((layer.image, alpha))
            pix = numpy_to_qpixmap(rgba).scaled(32, 32, Qt.AspectRatioMode.KeepAspectRatio, Qt.TransformationMode.SmoothTransformation)
            self.lbl_thumb.setPixmap(pix)
        else:
            self.lbl_thumb.setText("📁" if layer.layer_type == "group" else "T")
            self.lbl_thumb.setAlignment(Qt.AlignmentFlag.AlignCenter)

    def _toggle_visibility(self):
        visible = (self.btn_eye.text() == "🙈")
        self.btn_eye.setText("👁" if visible else "🙈")
        self.visibility_changed.emit(self.layer_id, visible)

    def _toggle_lock(self):
        locked = (self.btn_lock.text() == "🔓")
        self.btn_lock.setText("🔒" if locked else "🔓")
        self.lock_changed.emit(self.layer_id, locked)

    def start_rename(self):
        self.lbl_name.hide()
        self.txt_name.show()
        self.txt_name.setFocus()
        self.txt_name.selectAll()

    def _finish_rename(self):
        new_name = self.txt_name.text().strip()
        if new_name:
            self.lbl_name.setText(new_name)
            self.name_changed.emit(self.layer_id, new_name)
        self.txt_name.hide()
        self.lbl_name.show()


class LayerManagerPanel(QWidget):
    """
    Sidebar control panel for Layer Management.
    Supports Z-order drag & drop, visibility, locking, opacity, blend modes,
    grouping, duplicating, deleting, and reordering.
    """
    def __init__(self, parent=None):
        super().__init__(parent)
        self.document: ImageDocument = None
        self._is_updating_ui = False
        self._init_ui()

    def set_document(self, doc: ImageDocument):
        self.document = doc
        if self.document:
            self.document.add_change_listener(self.update_panel)
        self.update_panel()

    def _init_ui(self):
        layout = QVBoxLayout(self)
        layout.setContentsMargins(8, 8, 8, 8)
        layout.setSpacing(6)

        # Panel Header & Layer Actions Toolbar
        hdr_layout = QHBoxLayout()
        lbl_title = QLabel("Layers", self)
        lbl_title.setStyleSheet("font-size: 14px; font-weight: bold;")
        hdr_layout.addWidget(lbl_title, stretch=1)

        self.btn_add = QToolButton(self)
        self.btn_add.setText("➕")
        self.btn_add.setToolTip("Import New Image Layer")
        self.btn_add.setFixedSize(26, 26)

        self.btn_group = QToolButton(self)
        self.btn_group.setText("📁")
        self.btn_group.setToolTip("Group Selected Layers")
        self.btn_group.setFixedSize(26, 26)

        self.btn_dup = QToolButton(self)
        self.btn_dup.setText("📋")
        self.btn_dup.setToolTip("Duplicate Layer (Ctrl+D)")
        self.btn_dup.setFixedSize(26, 26)

        self.btn_del = QToolButton(self)
        self.btn_del.setText("🗑")
        self.btn_del.setToolTip("Delete Layer (Del)")
        self.btn_del.setFixedSize(26, 26)

        hdr_layout.addWidget(self.btn_add)
        hdr_layout.addWidget(self.btn_group)
        hdr_layout.addWidget(self.btn_dup)
        hdr_layout.addWidget(self.btn_del)
        layout.addLayout(hdr_layout)

        # Z-Order Quick Buttons (Top, Up, Down, Bottom)
        z_layout = QHBoxLayout()
        z_layout.setSpacing(4)
        self.btn_top = QPushButton("Top ⏫", self)
        self.btn_up = QPushButton("Up ▲", self)
        self.btn_down = QPushButton("Down ▼", self)
        self.btn_bot = QPushButton("Bottom ⏬", self)

        for btn in [self.btn_top, self.btn_up, self.btn_down, self.btn_bot]:
            btn.setFixedHeight(24)
            btn.setStyleSheet("font-size: 11px; padding: 2px 4px;")
            z_layout.addWidget(btn)
        layout.addLayout(z_layout)

        # Layer Stack List Widget (Drag & Drop Reordering)
        self.layer_list = QListWidget(self)
        self.layer_list.setSelectionMode(QAbstractItemView.SelectionMode.ExtendedSelection)
        self.layer_list.setDragDropMode(QAbstractItemView.DragDropMode.InternalMove)
        self.layer_list.setStyleSheet("""
            QListWidget { background: #1e1e1e; border: 1px solid #3a3a3a; border-radius: 4px; }
            QListWidget::item { padding: 2px; border-bottom: 1px solid #2a2a2a; }
            QListWidget::item:selected { background: #005fb8; border-radius: 4px; }
        """)
        self.layer_list.itemSelectionChanged.connect(self._on_selection_changed)
        self.layer_list.model().rowsMoved.connect(self._on_rows_moved)
        self.layer_list.itemDoubleClicked.connect(self._on_item_double_clicked)
        self.layer_list.setContextMenuPolicy(Qt.ContextMenuPolicy.CustomContextMenu)
        self.layer_list.customContextMenuRequested.connect(self._show_context_menu)
        layout.addWidget(self.layer_list, stretch=1)

        # Opacity & Blend Mode Section
        opts_layout = QVBoxLayout()
        opts_layout.setSpacing(4)

        # Opacity Slider
        opac_hdr = QHBoxLayout()
        lbl_opac = QLabel("Opacity:", self)
        self.lbl_opac_val = QLabel("100%", self)
        opac_hdr.addWidget(lbl_opac)
        opac_hdr.addStretch(1)
        opac_hdr.addWidget(self.lbl_opac_val)
        opts_layout.addLayout(opac_hdr)

        self.slider_opacity = QSlider(Qt.Orientation.Horizontal, self)
        self.slider_opacity.setRange(0, 100)
        self.slider_opacity.setValue(100)
        self.slider_opacity.valueChanged.connect(self._on_opacity_changed)
        opts_layout.addWidget(self.slider_opacity)

        # Blend Mode Dropdown
        bm_hdr = QHBoxLayout()
        lbl_bm = QLabel("Blend Mode:", self)
        self.cmb_blend_mode = QComboBox(self)
        self.cmb_blend_mode.addItems(["Normal", "Multiply", "Screen", "Overlay", "Darken", "Lighten", "Add", "Difference"])
        self.cmb_blend_mode.currentTextChanged.connect(self._on_blend_mode_changed)
        bm_hdr.addWidget(lbl_bm)
        bm_hdr.addWidget(self.cmb_blend_mode, stretch=1)
        opts_layout.addLayout(bm_hdr)

        layout.addLayout(opts_layout)

        # Signal connections for buttons
        self.btn_group.clicked.connect(self.action_group)
        self.btn_dup.clicked.connect(self.action_duplicate)
        self.btn_del.clicked.connect(self.action_delete)
        self.btn_top.clicked.connect(lambda: self.document and self.document.move_layer_top())
        self.btn_up.clicked.connect(lambda: self.document and self.document.move_layer_up())
        self.btn_down.clicked.connect(lambda: self.document and self.document.move_layer_down())
        self.btn_bot.clicked.connect(lambda: self.document and self.document.move_layer_bottom())

    def update_panel(self):
        """Refreshes layer list items and selection state from Document model efficiently."""
        if not self.document or self._is_updating_ui:
            return

        self._is_updating_ui = True
        try:
            doc_layer_ids_reversed = [lyr.id for lyr in reversed(self.document.layers)]
            current_item_ids = [self.layer_list.item(i).data(Qt.ItemDataRole.UserRole) for i in range(self.layer_list.count())]

            # Only rebuild list items if layer structure/order/count changed
            if doc_layer_ids_reversed != current_item_ids:
                self.layer_list.clear()

                # Render items in reverse order (Top layer at top of QListWidget)
                for lyr in reversed(self.document.layers):
                    item = QListWidgetItem(self.layer_list)
                    item.setSizeHint(QSize(0, 36))
                    item.setData(Qt.ItemDataRole.UserRole, lyr.id)

                    item_widget = LayerItemWidget(lyr)
                    item_widget.visibility_changed.connect(self._on_item_visibility_changed)
                    item_widget.lock_changed.connect(self._on_item_lock_changed)
                    item_widget.name_changed.connect(self._on_item_name_changed)

                    self.layer_list.setItemWidget(item, item_widget)

            # Update selection state
            for i in range(self.layer_list.count()):
                item = self.layer_list.item(i)
                lid = item.data(Qt.ItemDataRole.UserRole)
                item.setSelected(lid in self.document.active_layer_ids)

            # Update Opacity & Blend Mode from active layer
            active = self.document.active_layer
            if active:
                self.slider_opacity.blockSignals(True)
                self.slider_opacity.setValue(int(active.opacity * 100))
                self.lbl_opac_val.setText(f"{int(active.opacity * 100)}%")
                self.slider_opacity.blockSignals(False)

                self.cmb_blend_mode.blockSignals(True)
                self.cmb_blend_mode.setCurrentText(active.blend_mode)
                self.cmb_blend_mode.blockSignals(False)
        finally:
            self._is_updating_ui = False


    def _on_selection_changed(self):
        if self._is_updating_ui or not self.document:
            return
        selected_ids = []
        for item in self.layer_list.selectedItems():
            lid = item.data(Qt.ItemDataRole.UserRole)
            if lid:
                selected_ids.append(lid)
        if selected_ids:
            self.document.active_layer_ids = selected_ids
            self.document.notify_changed()

    def _on_rows_moved(self, parent, start, end, destination, row):
        """Translates list widget drag-and-drop move into document Z-order reordering."""
        if self._is_updating_ui or not self.document:
            return
        # QListWidget index 0 = top of list = last index in document.layers
        total = self.layer_list.count()
        new_layers = []
        for i in range(total - 1, -1, -1):
            item = self.layer_list.item(i)
            lid = item.data(Qt.ItemDataRole.UserRole)
            lyr = self.document.get_layer_by_id(lid)
            if lyr:
                new_layers.append(lyr)
        self.document.layers = new_layers
        self.document.notify_changed()

    def _on_item_double_clicked(self, item: QListWidgetItem):
        widget = self.layer_list.itemWidget(item)
        if isinstance(widget, LayerItemWidget):
            widget.start_rename()

    def _on_item_visibility_changed(self, layer_id: str, visible: bool):
        lyr = self.document.get_layer_by_id(layer_id)
        if lyr:
            lyr.visible = visible
            self.document.notify_changed()

    def _on_item_lock_changed(self, layer_id: str, locked: bool):
        lyr = self.document.get_layer_by_id(layer_id)
        if lyr:
            lyr.locked = locked
            self.document.notify_changed()

    def _on_item_name_changed(self, layer_id: str, new_name: str):
        lyr = self.document.get_layer_by_id(layer_id)
        if lyr:
            lyr.name = new_name
            self.document.notify_changed()

    def _on_opacity_changed(self, val: int):
        if self.document and self.document.active_layer:
            self.lbl_opac_val.setText(f"{val}%")
            self.document.active_layer.opacity = val / 100.0
            self.document.notify_changed()

    def _on_blend_mode_changed(self, mode: str):
        if self.document and self.document.active_layer:
            self.document.active_layer.blend_mode = mode
            self.document.notify_changed()

    def action_duplicate(self):
        if self.document:
            self.document.duplicate_layers()

    def action_delete(self):
        if self.document and self.document.active_layer_ids:
            self.document.remove_layers(list(self.document.active_layer_ids))

    def action_group(self):
        if self.document and len(self.document.active_layer_ids) >= 1:
            self.document.group_layers(list(self.document.active_layer_ids))

    def _show_context_menu(self, pos):
        item = self.layer_list.itemAt(pos)
        if not item or not self.document:
            return
        lid = item.data(Qt.ItemDataRole.UserRole)
        lyr = self.document.get_layer_by_id(lid)
        if not lyr:
            return

        menu = QMenu(self)
        act_rename = menu.addAction("✏ Rename")
        act_dup = menu.addAction("📋 Duplicate")
        act_del = menu.addAction("🗑 Delete")
        menu.addSeparator()
        act_top = menu.addAction("⏫ Move to Top")
        act_up = menu.addAction("▲ Move Up")
        act_down = menu.addAction("▼ Move Down")
        act_bot = menu.addAction("⏬ Move to Bottom")
        menu.addSeparator()
        act_group = menu.addAction("📁 Group Selected")
        act_ungroup = menu.addAction("📂 Ungroup")

        selected_action = menu.exec_(self.layer_list.mapToGlobal(pos))
        if selected_action == act_rename:
            widget = self.layer_list.itemWidget(item)
            if isinstance(widget, LayerItemWidget):
                widget.start_rename()
        elif selected_action == act_dup:
            self.document.duplicate_layers([lid])
        elif selected_action == act_del:
            self.document.remove_layers([lid])
        elif selected_action == act_top:
            self.document.move_layer_top(lid)
        elif selected_action == act_up:
            self.document.move_layer_up(lid)
        elif selected_action == act_down:
            self.document.move_layer_down(lid)
        elif selected_action == act_bot:
            self.document.move_layer_bottom(lid)
        elif selected_action == act_group:
            self.action_group()
        elif selected_action == act_ungroup:
            self.document.ungroup_layer(lid)

File: src/ui/panels/LayerPanel.cpp
#include "ui/panels/LayerPanel.h"
#include "processing/Compositor.h"
#include "utils/ImageUtils.h"
#include "utils/Logger.h"
#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QMenu>
#include <algorithm>

namespace ImageCut {
namespace UI {

// LayerItemWidget
LayerItemWidget::LayerItemWidget(std::shared_ptr<Core::Layer> layer, QWidget* parent)
    : QWidget(parent), m_layerId(layer->id)
{
    QHBoxLayout* layout = new QHBoxLayout(this);
    layout->setContentsMargins(4, 2, 4, 2);
    layout->setSpacing(6);

    m_btnEye = new QToolButton(this);
    m_btnEye->setText(layer->visible ? "👁" : "🙈");
    m_btnEye->setFixedSize(26, 26);
    m_btnEye->setStyleSheet("QToolButton { border: none; background: transparent; font-size: 14px; }");
    connect(m_btnEye, &QToolButton::clicked, [this]() {
        bool vis = (m_btnEye->text() == "🙈");
        m_btnEye->setText(vis ? "👁" : "🙈");
        emit visibilityChanged(m_layerId, vis);
    });
    layout->addWidget(m_btnEye);

    m_btnLock = new QToolButton(this);
    m_btnLock->setText(layer->locked ? "🔒" : "🔓");
    m_btnLock->setFixedSize(24, 24);
    m_btnLock->setStyleSheet("QToolButton { border: none; background: transparent; font-size: 12px; }");
    connect(m_btnLock, &QToolButton::clicked, [this]() {
        bool lock = (m_btnLock->text() == "🔓");
        m_btnLock->setText(lock ? "🔒" : "🔓");
        emit lockChanged(m_layerId, lock);
    });
    layout->addWidget(m_btnLock);

    m_lblThumb = new QLabel(this);
    m_lblThumb->setFixedSize(32, 32);
    m_lblThumb->setStyleSheet("QLabel { background: #2b2b2b; border: 1px solid #444; border-radius: 4px; }");
    updateThumbnail(layer);
    layout->addWidget(m_lblThumb);

    m_lblName = new QLabel(layer->name, this);
    m_lblName->setStyleSheet("QLabel { font-weight: bold; color: #e0e0e0; }");
    layout->addWidget(m_lblName, 1);

    m_txtName = new QLineEdit(layer->name, this);
    m_txtName->hide();
    connect(m_txtName, &QLineEdit::editingFinished, [this]() {
        QString newName = m_txtName->text().trimmed();
        if (!newName.isEmpty()) {
            m_lblName->setText(newName);
            emit nameChanged(m_layerId, newName);
        }
        m_txtName->hide();
        m_lblName->show();
    });
    layout->addWidget(m_txtName, 1);
}

void LayerItemWidget::updateThumbnail(std::shared_ptr<Core::Layer> layer) {
    if (!layer || layer->image.empty()) {
        m_lblThumb->setText(layer && layer->layerType == "group" ? "📁" : "T");
        m_lblThumb->setAlignment(Qt::AlignCenter);
        return;
    }

    cv::Mat renderMat = Processing::Compositor::renderSingleLayer(*layer);
    if (!renderMat.empty()) {
        QPixmap pix = Utils::ImageUtils::matToQPixmap(renderMat).scaled(32, 32, Qt::KeepAspectRatio, Qt::SmoothTransformation);
        m_lblThumb->setPixmap(pix);
    } else {
        QPixmap pix = Utils::ImageUtils::matToQPixmap(layer->image).scaled(32, 32, Qt::KeepAspectRatio, Qt::SmoothTransformation);
        m_lblThumb->setPixmap(pix);
    }
}

void LayerItemWidget::startRename() {
    m_lblName->hide();
    m_txtName->show();
    m_txtName->setFocus();
    m_txtName->selectAll();
}

// LayerManagerPanel
LayerManagerPanel::LayerManagerPanel(QWidget* parent)
    : QWidget(parent)
{
    initUi();
}

void LayerManagerPanel::setDocument(std::shared_ptr<Core::ImageDocument> doc) {
    m_document = doc;
    if (m_document) {
        m_document->addChangeListener([this]() {
            updatePanel();
        });
    }
    updatePanel();
}

void LayerManagerPanel::initUi() {
    QVBoxLayout* layout = new QVBoxLayout(this);
    layout->setContentsMargins(8, 8, 8, 8);
    layout->setSpacing(8);

    QHBoxLayout* topHbox = new QHBoxLayout();
    topHbox->setSpacing(4);

    btnAdd = new QToolButton(this); btnAdd->setText("➕ Add");
    m_btnDup = new QToolButton(this); m_btnDup->setText("📋 Dup");
    m_btnDel = new QToolButton(this); m_btnDel->setText("🗑 Del");
    m_btnGroup = new QToolButton(this); m_btnGroup->setText("📁 Group");

    topHbox->addWidget(btnAdd);
    topHbox->addWidget(m_btnDup);
    topHbox->addWidget(m_btnDel);
    topHbox->addWidget(m_btnGroup);
    layout->addLayout(topHbox);

    QHBoxLayout* reorderHbox = new QHBoxLayout();
    reorderHbox->setSpacing(4);

    m_btnTop = new QPushButton("Top 🔝", this);
    m_btnUp = new QPushButton("Up 🔺", this);
    m_btnDown = new QPushButton("Down 🔻", this);
    m_btnBot = new QPushButton("Bottom ⏹", this);

    reorderHbox->addWidget(m_btnTop);
    reorderHbox->addWidget(m_btnUp);
    reorderHbox->addWidget(m_btnDown);
    reorderHbox->addWidget(m_btnBot);
    layout->addLayout(reorderHbox);

    m_layerList = new QListWidget(this);
    m_layerList->setSelectionMode(QAbstractItemView::ExtendedSelection);
    m_layerList->setContextMenuPolicy(Qt::CustomContextMenu);
    connect(m_layerList, &QListWidget::itemSelectionChanged, this, &LayerManagerPanel::onSelectionChanged);
    connect(m_layerList, &QListWidget::itemDoubleClicked, this, &LayerManagerPanel::onItemDoubleClicked);
    connect(m_layerList, &QListWidget::customContextMenuRequested, this, &LayerManagerPanel::showContextMenu);
    layout->addWidget(m_layerList, 1);

    QVBoxLayout* optsLayout = new QVBoxLayout();
    optsLayout->setSpacing(6);

    QHBoxLayout* opacHdr = new QHBoxLayout();
    opacHdr->addWidget(new QLabel("Opacity:", this));
    m_lblOpacVal = new QLabel("100%", this);
    opacHdr->addStretch(1);
    opacHdr->addWidget(m_lblOpacVal);
    optsLayout->addLayout(opacHdr);

    m_sliderOpacity = new QSlider(Qt::Horizontal, this);
    m_sliderOpacity->setRange(0, 100);
    m_sliderOpacity->setValue(100);
    connect(m_sliderOpacity, &QSlider::valueChanged, [this](int val) {
        if (m_document && m_document->getActiveLayer()) {
            m_lblOpacVal->setText(QString("%1%").arg(val));
            m_document->getActiveLayer()->opacity = val / 100.0;
            m_document->notifyChanged();
        }
    });
    optsLayout->addWidget(m_sliderOpacity);

    QHBoxLayout* bmHdr = new QHBoxLayout();
    bmHdr->addWidget(new QLabel("Blend Mode:", this));
    m_cmbBlendMode = new QComboBox(this);
    m_cmbBlendMode->addItems({ "Normal", "Multiply", "Screen", "Overlay", "Darken", "Lighten", "Add", "Difference" });
    connect(m_cmbBlendMode, &QComboBox::currentTextChanged, [this](const QString& mode) {
        if (m_document && m_document->getActiveLayer()) {
            m_document->getActiveLayer()->blendMode = mode;
            m_document->notifyChanged();
        }
    });
    bmHdr->addWidget(m_cmbBlendMode, 1);
    optsLayout->addLayout(bmHdr);

    layout->addLayout(optsLayout);

    connect(m_btnDup, &QToolButton::clicked, [this]() { if (m_document) m_document->duplicateLayers(); });
    connect(m_btnDel, &QToolButton::clicked, [this]() { if (m_document) m_document->removeLayers(m_document->activeLayerIds); });
    connect(m_btnGroup, &QToolButton::clicked, [this]() { if (m_document) m_document->groupLayers(m_document->activeLayerIds); });

    connect(m_btnTop, &QPushButton::clicked, [this]() { if (m_document) m_document->moveLayerTop(); });
    connect(m_btnUp, &QPushButton::clicked, [this]() { if (m_document) m_document->moveLayerUp(); });
    connect(m_btnDown, &QPushButton::clicked, [this]() { if (m_document) m_document->moveLayerDown(); });
    connect(m_btnBot, &QPushButton::clicked, [this]() { if (m_document) m_document->moveLayerBottom(); });
}

void LayerManagerPanel::updatePanel() {
    if (!m_document || m_isUpdatingUi) return;

    m_isUpdatingUi = true;
    m_layerList->blockSignals(true);
    try {
        m_layerList->clear();

        for (auto it = m_document->layers.rbegin(); it != m_document->layers.rend(); ++it) {
            auto lyr = *it;
            QListWidgetItem* item = new QListWidgetItem(m_layerList);
            item->setSizeHint(QSize(0, 36));
            item->setData(Qt::UserRole, lyr->id);

            LayerItemWidget* widget = new LayerItemWidget(lyr, this);
            connect(widget, &LayerItemWidget::visibilityChanged, [this](const QString& lid, bool vis) {
                auto l = m_document->getLayerById(lid);
                if (l) { l->visible = vis; m_document->notifyChanged(); }
            });
            connect(widget, &LayerItemWidget::lockChanged, [this](const QString& lid, bool lock) {
                auto l = m_document->getLayerById(lid);
                if (l) { l->locked = lock; m_document->notifyChanged(); }
            });
            connect(widget, &LayerItemWidget::nameChanged, [this](const QString& lid, const QString& name) {
                auto l = m_document->getLayerById(lid);
                if (l) { l->name = name; m_document->notifyChanged(); }
            });

            m_layerList->setItemWidget(item, widget);
            if (std::find(m_document->activeLayerIds.begin(), m_document->activeLayerIds.end(), lyr->id) != m_document->activeLayerIds.end()) {
                item->setSelected(true);
            }
        }

        auto active = m_document->getActiveLayer();
        if (active) {
            m_sliderOpacity->blockSignals(true);
            m_sliderOpacity->setValue(static_cast<int>(active->opacity * 100));
            m_lblOpacVal->setText(QString("%1%").arg(static_cast<int>(active->opacity * 100)));
            m_sliderOpacity->blockSignals(false);

            m_cmbBlendMode->blockSignals(true);
            m_cmbBlendMode->setCurrentText(active->blendMode);
            m_cmbBlendMode->blockSignals(false);
        }
    } catch (...) {}
    m_layerList->blockSignals(false);
    m_isUpdatingUi = false;
}

void LayerManagerPanel::onSelectionChanged() {
    if (m_isUpdatingUi || !m_document) return;
    std::vector<QString> selIds;
    for (auto item : m_layerList->selectedItems()) {
        QString lid = item->data(Qt::UserRole).toString();
        if (!lid.isEmpty()) selIds.push_back(lid);
    }
    if (!selIds.empty()) {
        m_document->activeLayerIds = selIds;
        m_document->notifyChanged();
    }
}

void LayerManagerPanel::onItemDoubleClicked(QListWidgetItem* item) {
    QWidget* w = m_layerList->itemWidget(item);
    LayerItemWidget* widget = qobject_cast<LayerItemWidget*>(w);
    if (widget) {
        widget->startRename();
    }
}

void LayerManagerPanel::showContextMenu(const QPoint& pos) {
    QListWidgetItem* item = m_layerList->itemAt(pos);
    if (!item || !m_document) return;

    QString lid = item->data(Qt::UserRole).toString();
    auto lyr = m_document->getLayerById(lid);
    if (!lyr) return;

    QMenu menu(this);
    QAction* actRename = menu.addAction("✏ Rename");
    QAction* actDup = menu.addAction("📋 Duplicate");
    QAction* actDel = menu.addAction("🗑 Delete");
    menu.addSeparator();
    QAction* actTop = menu.addAction("⏫ Move to Top");
    QAction* actUp = menu.addAction("▲ Move Up");
    QAction* actDown = menu.addAction("▼ Move Down");
    QAction* actBot = menu.addAction("⏬ Move to Bottom");
    menu.addSeparator();
    QAction* actGroup = menu.addAction("📁 Group Selected");
    QAction* actUngroup = menu.addAction("📂 Ungroup");

    QAction* selectedAction = menu.exec(m_layerList->mapToGlobal(pos));
    if (selectedAction == actRename) {
        QWidget* w = m_layerList->itemWidget(item);
        LayerItemWidget* widget = qobject_cast<LayerItemWidget*>(w);
        if (widget) widget->startRename();
    } else if (selectedAction == actDup) {
        m_document->duplicateLayers({ lid });
    } else if (selectedAction == actDel) {
        m_document->removeLayers({ lid });
    } else if (selectedAction == actTop) {
        m_document->moveLayerTop(lid);
    } else if (selectedAction == actUp) {
        m_document->moveLayerUp(lid);
    } else if (selectedAction == actDown) {
        m_document->moveLayerDown(lid);
    } else if (selectedAction == actBot) {
        m_document->moveLayerBottom(lid);
    } else if (selectedAction == actGroup) {
        m_document->groupLayers(m_document->activeLayerIds);
    } else if (selectedAction == actUngroup) {
        m_document->ungroupLayer(lid);
    }
}

} // namespace UI
} // namespace ImageCut
